pkg/router/template/router.go

```diff
@@ -1286,6 +1286,16 @@ func (r *templateRouter) getActiveEndpoints(serviceUnits map[ServiceUnitKey]int3
 // The above assumes roundRobin scheduling.
 func (r *templateRouter) calculateServiceWeights(serviceUnits map[ServiceUnitKey]int32) map[ServiceUnitKey]int32 {
 	serviceUnitNames := make(map[ServiceUnitKey]int32)
+
+	// If there is only 1 service unit, then always set the weight 1 for all the endpoints.
+	// Scaling the weight to 256 is redundant and causes haproxy to allocate more memory on startup.
+	if len(serviceUnits) == 1 {
+		for key := range serviceUnits {
+			serviceUnitNames[key] = 1
+		}
```

Comment:
What if the service unit has 0 endpoints?  We should check for that case:
```suggestion
		for key := range serviceUnits {
			if r.numberOfEndpoints(key) > 0 {
				serviceUnitNames[key] = 1
			}
		}
```
Granted, omitting the `r.numberOfEndpoints(key) > 0` check might not change the ultimate result from config template:
```
        {{- range $serviceUnitName, $weight := $cfg.ServiceUnitNames }}
          {{- if ge $weight 0 }}{{/* weight=0 is reasonable to keep existing connections to backends with cookies as we can see the HTTP headers */}}
            {{- with $serviceUnit := index $.ServiceUnits $serviceUnitName }}
              {{- range $idx, $endpoint := processEndpointsForAlias $cfg $serviceUnit (env "ROUTER_BACKEND_PROCESS_ENDPOINTS" "") }}
                {{/* [actual HAProxy config stuff here] */}}
            {{- end }}{{/* end range processEndpointsForAlias */}}
          {{- end }}{{/* end get serviceUnit from its name */}}
        {{- end }}{{/* end range over serviceUnitNames */}}

```
In the template, the effect is the same whether `range $serviceUnitName, $weight := $cfg.ServiceUnitNames` iterates 0 times or whether `range $idx, $endpoint := processEndpointsForAlias $cfg $serviceUnit (env "ROUTER_BACKEND_PROCESS_ENDPOINTS" "")` iterates 0 times.  However, it could cause problems for the dynamic config manager, which has logic like this:
```go
		// As the endpoints have changed, recalculate the weights.
		newWeights := r.calculateServiceWeights(cfg.ServiceUnits)

		// Get the weight for this service unit.
		weight, ok := newWeights[id]
		if !ok {
			weight = 0
		}
```
Anyway, from a strict correctness perspective, I believe `calculateServiceWeights` needs the `r.numberOfEndpoints(key) > 0` check.  

Please add a unit test case to ensure we don't regress:
```diff
diff --git a/pkg/router/template/router.go b/pkg/router/template/router.go
index 1641b0c7..c6a65a79 100644
--- a/pkg/router/template/router.go
+++ b/pkg/router/template/router.go
@@ -1287,6 +1287,20 @@ func (r *templateRouter) getActiveEndpoints(serviceUnits map[ServiceUnitKey]int3
 func (r *templateRouter) calculateServiceWeights(serviceUnits map[ServiceUnitKey]int32) map[ServiceUnitKey]int32 {
 	serviceUnitNames := make(map[ServiceUnitKey]int32)
 	// portion of service weight for each endpoint
+
+
+	// If there is only 1 service unit, then always set the weight 1 for all the endpoints.
+	// Scaling the weight to 256 is redundant and causes haproxy to allocate more memory on startup.
+	if len(serviceUnits) == 1 {
+		for key := range serviceUnits {
+			if r.numberOfEndpoints(key) > 0 {
+				serviceUnitNames[key] = 1
+			}
+		}
+		return serviceUnitNames
+	}
+
+
 	epWeight := make(map[ServiceUnitKey]float32)
 	// maximum endpoint weight
 	var maxEpWeight float32 = 0.0
diff --git a/pkg/router/template/router_test.go b/pkg/router/template/router_test.go
index 73e77f55..1d88c983 100644
--- a/pkg/router/template/router_test.go
+++ b/pkg/router/template/router_test.go
@@ -873,6 +873,16 @@ func TestCalculateServiceWeights(t *testing.T) {
 		serviceWeights  map[ServiceUnitKey]int32
 		expectedWeights map[ServiceUnitKey]int32
 	}{
+		{
+			name: "service with no endpoint",
+			serviceUnits: map[ServiceUnitKey][]Endpoint{
+				suKey1: {},
+			},
+			serviceWeights: map[ServiceUnitKey]int32{
+				suKey1: 100,
+			},
+			expectedWeights: map[ServiceUnitKey]int32{},
+		},
 		{
 			name: "equally weighted services with same number of endpoints",
 			serviceUnits: map[ServiceUnitKey][]Endpoint{
```

Reply:
Good point, didn't think about a single service with no endpoints and exposing that datapath. Agreed it doesn't functionally changing anything, but keep our code clean for the future.

Done.